spec/notification_center_spec.rb

```diff
@@ -0,0 +1,484 @@
+  let(:config_body) { OptimizelySpec::VALID_CONFIG_BODY }
+  let(:config_body_JSON) { OptimizelySpec::VALID_CONFIG_BODY_JSON }
+  let(:error_handler) { Optimizely::NoOpErrorHandler.new }
+  let(:logger) { Optimizely::NoOpLogger.new }
+  let(:notification_center) { Optimizely::NotificationCenter.new(spy_logger, error_handler)}
+  
+  before(:context) do
+    class CallBack
+      def call(args)
+        args
+      end
+    end
+
+    @callback = CallBack.new
+    @callback_reference = @callback.method(:call)
+  end
+  
+  describe '#Notification center' do
+    
+    describe 'test add notification with invalid params' do
+      it 'should log and return nil if notification type is empty' do
+        expect(notification_center.add_notification_listener(
+         nil,
+         @callback_reference
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Notification type can not be empty.')
+      end
+  
+      it 'should log and return nil if notification callback is empty' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         nil
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Callback can not be empty.')
+      end
+      
+      it 'should log and return nil if invalid notification type given' do
+        expect(notification_center.add_notification_listener(
+         'Test notification type',
+         @callback_reference
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+                               .with(Logger::ERROR, 'Invalid notification type.')
+      end
+      
+      it 'should log and return nil if invalid callable given' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         'Invalid callback!'
+        )).to eq(nil)
+        expect(spy_logger).to have_received(:log).once
+           .with(Logger::ERROR, 'Invalid notification callback given.')
+      end
+    end
+    
+    describe 'test add notification with valid type and callback' do
+      it 'should add, and return notification ID when a plain function is passed as an argument ' do
+        expect(notification_center.add_notification_listener(
+         Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE],
+         @callback_reference
+        )).to eq(1)
+        # verifies that one notification is added
+        expect(notification_center.notifications[Optimizely::NotificationCenter::NOTIFICATION_TYPES[:ACTIVATE]].length)
+         .to eq(1)
+      end
+    end
+    
+    describe 'test add notification for multiple notification types' do
+      it 'should add, return notification ID and validate callbacks when a valid callback is added for each notification type ' do
```

Comment:
Do not validate callback here. See comment below. You just have to make sure that when you send a valid callback for each type. It is added in the list